attestation/jsonattestor/json.go

```diff
@@ -0,0 +1,213 @@
+package jsonattestor
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/cyberphone/json-canonicalization/go/src/webpki.org/jsoncanonicalizer"
+	"github.com/ghodss/yaml"
+	"github.com/itchyny/gojq"
+
+	"github.com/in-toto/go-witness/attestation"
+	"github.com/in-toto/go-witness/cryptoutil"
+	"github.com/in-toto/go-witness/log"
+	"github.com/in-toto/go-witness/registry"
+	"github.com/invopop/jsonschema"
+)
+
+const (
+	Name    = "json"
+	Type    = "https://witness.dev/attestations/json/v0.1"
+	RunType = attestation.PostProductRunType
+)
+
+// Attestor processes a JSON/YAML file, storing canonical JSON and user queries.
+type Attestor struct {
+	CanonicalJSON  []byte
+	SubjectQueries map[string]string
+	subjects       map[string]cryptoutil.DigestSet
+}
+
+var (
+	_ attestation.Attestor  = &Attestor{}
+	_ attestation.Subjecter = &Attestor{}
+)
+
+func init() {
+	attestation.RegisterAttestation(
+		Name,
+		Type,
+		RunType,
+		func() attestation.Attestor { return New() },
+		registry.StringConfigOption(
+			"subject-queries",
+			"subjectName=jq expression pairs (semicolon-separated). Example: 'foo=.foo;bar=.nums[]'",
+			"",
+			func(a attestation.Attestor, sqStr string) (attestation.Attestor, error) {
+				att, ok := a.(*Attestor)
+				if !ok {
+					return a, fmt.Errorf("unexpected attestor type: %T", a)
+				}
+				sqMap, err := ParseSubjectQueries(sqStr)
+				if err != nil {
+					return a, fmt.Errorf("parseSubjectQueries: %w", err)
+				}
+				att.SubjectQueries = sqMap
+				return att, nil
+			},
+		),
+	)
+}
+
+func New() *Attestor {
+	return &Attestor{
+		subjects:       make(map[string]cryptoutil.DigestSet),
+		SubjectQueries: make(map[string]string),
+	}
+}
+
+func ParseSubjectQueries(sqStr string) (map[string]string, error) {
+	out := make(map[string]string)
+	s := strings.TrimSpace(sqStr)
+	if s == "" {
+		return out, nil
+	}
+	pairs := strings.Split(s, ";")
+	for _, pair := range pairs {
+		pair = strings.TrimSpace(pair)
+		if pair == "" {
+			continue
+		}
+		parts := strings.SplitN(pair, "=", 2)
+		if len(parts) != 2 {
```

Comment:
Correct me if i'm wrong, but given you've set `strings.SplitN(..., 2)`, there will *always* be 2 parts right? It will just stop splitting when it reaches 2 parts. E.g., `foo=.foo=bar=.bar` will return:
- `foo`
- `.foo=bar=.bar`

We could fix this by using `strings.Split`, and we will then always fail if the length isn't 2.

Reply:
<img width="508" alt="image" src="https://github.com/user-attachments/assets/6ceaf403-22c5-457b-9f0c-3ccd5e4d3faf" />

Reply:
We’re using strings.SplitN(pair, "=", 2) because the jq expression may contain additional = signs, e.g. .foo=bar, or more complex expressions with = in them. In that scenario, we only want to split the string on the first =, treating everything after it as the jq expression. If we used a plain strings.Split(pair, "=") and enforced exactly two parts, then any extra = within the expression would cause an error.